middle_end/flambda2/simplify/simplify_let_cont_expr.ml

```diff
@@ -728,19 +728,56 @@ let simplify_non_recursive_let_cont ~simplify_expr dacc non_rec ~down_to_up =
       (simplify_non_recursive_let_cont_stage0 ~simplify_expr dacc non_rec
          ~down_to_up)
 
+type recursive_let_cont_handlers_element =
+  | Non_recursive_handler of Continuation.t * RE.Continuation_handler.t
+  | Recursive_handlers of RE.Continuation_handler.t Continuation.Map.t
+
 let rebuild_recursive_let_cont_handlers cont ~params ~original_cont_scope
-    cont_handler ~handler uacc ~after_rebuild =
+    cont_handler ~handler ~free_names_of_handler ~cost_metrics_of_handler
+    ~cont_uses_env_after_body uacc ~after_rebuild =
+  let is_recursive =
+    Continuation.Set.mem cont
+      (Name_occurrences.continuations_including_in_trap_actions
+         free_names_of_handler)
+  in
+  let is_inlinable =
+    if is_recursive
+    then false
+    else
+      match CUE.get_continuation_uses cont_uses_env_after_body cont with
+      | None -> false
+      | Some uses -> (
+        match Continuation_uses.get_uses uses with
+        | [use] -> (
+          match One_continuation_use.use_kind use with
+          | Inlinable -> true
+          | Non_inlinable _ -> false)
+        | _ -> false)
+  in
   let uacc =
     UA.map_uenv uacc ~f:(fun uenv ->
-        UE.add_non_inlinable_continuation uenv cont original_cont_scope ~params
-          ~handler:(Known handler))
+        if is_inlinable
+        then
+          UE.add_linearly_used_inlinable_continuation uenv cont
+            original_cont_scope ~params ~handler ~free_names_of_handler
+            ~cost_metrics_of_handler
+        else
+          UE.add_non_inlinable_continuation uenv cont original_cont_scope
+            ~params ~handler:(Known handler))
+  in
+  let handlers =
+    if is_recursive
+    then Recursive_handlers (Continuation.Map.singleton cont cont_handler)
+    else Non_recursive_handler (cont, cont_handler)
   in
-  let handlers = Continuation.Map.singleton cont cont_handler in
   after_rebuild handlers uacc
 
 let after_one_recursive_let_cont_handler_rebuilt cont ~original_cont_scope
     ~name_occurrences_subsequent_exprs ~after_rebuild cont_handler ~params
-    ~handler ~free_names_of_handler:_ ~cost_metrics_of_handler:_ uacc =
+    ~cont_uses_env_after_body ~handler ~free_names_of_handler
+    ~cost_metrics_of_handler uacc =
+  assert (
+    Name_occurrences.equal free_names_of_handler (UA.name_occurrences uacc));
```

Comment:
I think that assertion should be either removed or behind a `Flambda_features.check_invariants ()` test.